doc/config.js.md
>**[Return to the Masterportal documentation](doc.md)**

[TOC]

# config.js

The `config.js` contains Masterportal configuration not directly related to UI or layers. For example, paths to other configuration files belong here. This file is usually placed next to the `index.html` and `config.json` files.

In the following, all configuration options are described. For all configuration options of type `object`, further nested options are linked and described in detail after the main table. You may also refer to **[this config.js example file](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/portal/basic/config.js)**.

|Name|Required|Type|Default|Description|Example|
|----|--------|----|-------|-----------|-------|
|layerConf|yes|String||Path to the **[services.json](services.json.md)** file containing all available WMS layers and WFS feature types. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/services-internet.json"`|
|namedProjections|yes|String[]||Definition of the usable coordinate systems. See **[syntax definition](http://proj4js.org/#named-projections)** for details..|`[["EPSG:25832", "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"]]`|
|restConf|yes|String||Path to the **[rest-services.json](rest-services.json.md)** file describing further services, e.g. print service, WPS, CSW. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/rest-services-internet.json"`|
|styleConf|yes|String||Path to the **[style.json](style.json.md)** file describing vector layer (WFS) styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/style.json"`|
|addons|no|String[]|`[]`|List of names for custom modules. The modules are to be placed in the folder `/addons/`, with their entry points being defined in the `addonsConf.json`.|`["myAddon1", "myAddon2"]`|
|alerting|no|**[alerting](#markdown-header-alerting)**|`{"category": "alert-info", "isDismissable": true, "isConfirmable": false, "position": "top-center", "fadeOut": null}`|Overrides the alert module's default values.|{fadeOut: 6000}|
|cameraParameter|no|**[cameraParameter](#markdown-header-cameraparameter)**||Initial camera parameter||
|cesiumParameter|no|**[cesiumParameter](#markdown-header-cesiumparameter)**||Cesium flags||
|cswId|no|String|`"3"`|Reference to a CSW interface used to retrieve layer information. The ID will be resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`"my CSW-ID"`|
|defaultToolId|no|String|`"gfi"`|The tool with the given ID will be active when no other tool is active.|"filter"|
|layerSelector|no|**[layerSelector](#markdown-header-layerselector)**||Module to configure interactions with the layertree and the map, executed on a defined event.||
|featureViaURL|no|**[featureViaURL](#markdown-header-featureviaurl)**||Optional configuration for the URL parameter `featureViaURL`. See **[urlParameter](urlParameter.md)** for details. Implemented for treeTypes *light* and *custom*.||
|footer|no|**[footer](#markdown-header-footer)**||If set, a footer is shown and configured with this object.||
|gfiWindow|no|String|`"detached"`|_Deprecated in the next major release. Please use the attribute "Portalconfig.menu.tool.gfi.desktopType" of the **[config.json](#config.json.md)** instead._ Display type and attribute information for all layer types. **attached**: the attribute information window is opened at click position **detached**: the attribute information window is opened at the top right of the map; a marker is set to the click position.|`"attached"`|
|ignoredKeys|no|String[]|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|List of attribute names to be ignored for attribute information lists of all layer types. Only used with "gfiAttributes": "showAll".|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|
|infoJson|no|String|`"info.json"`|Path to the `info.json` file containing additional information on snippets. The path is relative to the *index.html*.|`"info.json"`|
|inputMap|no|Object|`{}`|If this object is set, and its field `setMarker` is set to true, the Masterportal is configured as input element for data. In that case, each click sets a Map Marker and communicates the coordinates via **[remoteInterface](remoteInterface.md)** in the chosen coordinate reference system.|`{setMarker: true, targetProjection: "EPSG:4326", setCenter: false}`|
|inputMap.setCenter|no|Boolean|`false`|Center on a marker after producing it?|`setCenter: true`|
|inputMap.setMarker|no|Boolean|`false`|Flag to activate the 'setMarker' functionality.|`setMarker: true`|
|inputMap.targetProjection|no|String|`"EPSG:25832"`|The target coordinate reference system. Coordinates will be translated to it before being communicated via **[remoteInterface](remoteInterface.md)**.|`targetprojection: "EPSG:4326"`|
|login|no|Object|`{}`|Activates the Login module. This displays a login button in the toolbar that allows login and logout with an IDM server, e.g. keycloak, as configured.||
|mapInteractions|no|**[mapInteractions](#markdown-header-mapInteractions)**||Overrides the ol map interactions. Provides further configuration possibilities for control behaviour and keyboardEventTarget.||
|mapMarker|no|**[mapMarker](#markdown-header-mapmarker)**||Overrides the map marker module's default values. Useful for 3D markers since OpenLayers's overlays can not be displayed in 3D mode. For this, the map marker has to be defined as vector layer.||
|metaDataCatalogueId|no|String|`"2"`|URL to the metadata catalog linked to in the layer information window. The ID is resolved to a service of the **[rest-services.json](rest-services.json.md)** file. Note: This attribute is only necessary, when no "show_doc_url" is configured in the metadata dataset in the **[services.json](services.json.md)**. The url can either be set globally (**[config.js](config.js.md)**) or layer-specific(**[services.json](services.json.md)**).|`"MetaDataCatalogueUrl"`|
|metadata|no|**[metadata](#markdown-header-metadata)**||Allows configuration of which metadata URLs are to be resolved via proxy.||
|mouseHover|no|**[mouseHover](#markdown-header-mousehover)**||Activates the MouseHover feature for vector layers, both WFS and GeoJSON. For per-layer configuration, see the **[config.json](config.json.md)**'s section *Themenconfig.Fachdaten.Layer*.|`true`|
|obliqueMap|no|Boolean|`false`|If set to `true`, an oblique map layer is created. An additional oblique layer must be defined.||
|portalConf|no|String|`"config.json"`|Path to the portal's `config.json` file. You may also enter a node; in that case the taken path is controlled by the urlParameter `config`.|Direct path: "../masterTree/config.json"; Node: "../../portal/master/". In the node scenario, a query parameter like `config=config.json` must exist in the URL.|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal is supposed to post messages to and receive messages from with the `postMessage` feature.|"http://localhost:8080"|
|proxyHost|no|String||Host name of a remote proxy with CORS configured to support the portal's domain, among others.|`"https://proxy.example.com"`|
|quickHelp|no|Object|`{}`|Activates the QuickHelp module. This displays a window containing help text for supported functions of the modules. Available for the layer tree (CustomTree), the search bar (Searchbar) and the routing tool (RoutingTool).||
|remoteInterface|no|**[remoteInterface](#markdown-header-remoteinterface)**||Optional remote interface configuration.||
|scaleLine|no|Boolean|`false`|Controls whether a scale line is displayed at the bottom right of the map. To activate scale selection (as a **ScaleSwitcher**) use the configuration parameter `isDisplayInFooter` in **[config.json](config.json)**.|`true`|
|simpleMap|no|Boolean|`false`|_Deprecated in the next major release. Please use the parameter `simpleMap` as part of the configuration of the `saveSelection` tool in the **[config.json](config.json.md)**._ Adds a SimpleMap URL to the `Save selection` dialogue. When calling this URL, the menu bar, layer tree, and map controls are deactivated. Not implemented for tree type *„light“*.|`false`|
|startingMap3D|bi|Boolean|`false`|Controls whether the map should start in 3D mode.||
|tree|no|**[tree](#tree)**||||
|uiStyle|no|String|`"default"`|Sets the control element layout. |`table`|
|wfsImgPath|no|String||Path to the folder holding images for the WFS styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/img/"`|
|wpsID|no|String|`""`|Reference to a WPS interface used in various modules. The ID is resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`""`|
|zoomToFeature|no|**[zoomToFeature](#markdown-header-zoomtofeature)**||_Deprecated in the next major release. Please use **[zoomTo](#markdown-header-zoomto)** instead._ Optional configuration of the URL query parameter `featureid`. For details, see **[urlParameter](urlParameter.md)**. ||
|zoomTo|no|**[zoomTo](#markdown-header-zoomto)**[]|Configuration for the URL query parameters `zoomToFeatureId` and `zoomToGeometry`.||
|layerInformation|no|**[layerInformation](#markdown-header-layerinformation)**||Configuration for the layerInformation window.||
|vuetify|no|String|undefined|Path to the optional instance of the vuetify UI library. e.g. portal or addon specific.|`addons/cosi/vuetify/index.js`|
|layerSequence|no|**[layerSequence](#markdown-header-layersequence)**||Configuration for layerSequence.||

***

## alerting

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|fetchBroadcastUrl|no|String|`false`|The alerting module will initially use a linked configuration file from this URL, if set.|
|localStorageDisplayedAlertsKey|no|String|`"displayedAlerts"`|Arbitrary key used to store information regarding the alerting module in the browser's local storage.|

***

## cameraParameter

Cesium Scene camera settings in 3D mode.
_Deprecated in the next major release. Please use **[cesiumParameter](#markdown-header-cesiumParameter)** instead._

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|
|altitude|no|Number||Camera's initial height in meters|

***

## cesiumParameter

Cesium Scene settings in 3D mode.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Scene.html?classFilter=scene)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|camera|no|**[camera](#markdown-header-cesiumParametercamera)**||Cesium Scene camera settings in 3D mode.|
|fog|no|**[fog](#markdown-header-cesiumParameterfog)**||Cesium Scene fog settings in 3D mode.|
|fxaa|no|Boolean|`true`|activates *fast approximate anti-aliasing*|
|globe|no|**[globe](#markdown-header-cesiumParameterglobe)**||Cesium Scene globe settings in 3D mode.|
|maximumScreenSpaceError|no|Number|`2.0`|Detail level in which terrain/raster tiles are fetched. 4/3 is the highest quality level.|
|tileCacheSize|no|Number|`100`|terrain/raster tile cache size|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    },
    "fog": {
        "enabled": true
    },
    "fxaa": true,
    "globe": {
        "enableLighting": true
    },
    "maximumScreenSpaceError": 2,
    "tileCacheSize": 20,
}
```
***

### cesiumParameter.camera

Cesium Scene camera settings in 3D mode.
The camera is defined by a position, orientation, and view frustum.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Camera.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|altitude|no|Number||Camera's initial height in meters|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|

**Example**

```json
{
    "camera": {
        "altitude": 127,
        "heading": -1.2502079000000208,
        "tilt": 45
    }
}
```
***

### cesiumParameter.fog

Cesium Scene fog settings in 3D mode.
Blends the atmosphere to geometry far from the camera for horizon views.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Fog.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|no|Boolean|`false`|True if fog is enabled.|

**Example**

```json
{
    "fog": {
        "enabled": true
    }
}
```
***

### cesiumParameter.globe

Cesium Scene globe settings in 3D mode.
The globe rendered in the scene, including its terrain and imagery layers.
For more attributes see **[Scene](https://cesium.com/learn/cesiumjs/ref-doc/Globe.html)**

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enableLighting|no|Boolean|`false`|Activates light effects on the map based on the sun's position.|

**Example**

```json
{
    "globe": {
        "enableLighting": true
    }
}
```
***

## footer

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|urls|no|**[urls](#markdown-header-footerurls)**[]||Array of URL configuration objects.|
|showVersion|no|Boolean|`false`|If `true`, the Masterportal version number is included in the footer.|
|footerInfo|no|**[footerInfo](#markdown-header-footerfooterInfo)**[]||Array of information configuration objects.|
|mobileFooterInfoToggler|no|Boolean|`false`|If `true`, it is possible to toggle the footer info in mobile version.|

***

### footer.urls
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|alias|no|String|`"Landesbetrieb Geoinformation und Vermessung"`|Link text for desktop applications|
|alias_mobil|no|String|`"LGV"`|Link text for mobile applications|
|bezeichnung|no|String|`"Kartographie und Gestaltung: "`|Link prefix|
|url|no|String||the URL the `alias` (or `alias_mobile`) text links to, e.g. `"https://example.com/"`. You may also link to mail creation with e.g. `"mailto:my_mail@example.com"`|
|toolModelId|no|String|`"SdpDownload"`|The id of a module to be opened on clicking the link. Do not include a URL for such cases.|

**Example:**

```json
{
    "footer": {
        "urls": [
            {
                "bezeichnung": "Cartography and design: ",
                "url": "https://geoinfo.hamburg.de/",
                "alias": "Landesbetrieb Geoniformation und Vermessung",
                "alias_mobil": "LGV"
            },
            {
                "bezeichnung": "",
                "url": "http://www.hamburg.de/bsu/timonline",
                "alias": "Map inconsistencies"
            },
            {
                "bezeichnung": "",
                "url": "",
                "alias": "SDP Download",
                "toolModelId": "sdpdownload"
            }
        ],
        "showVersion": true
    }
}
```

***

### footer.footerInfo
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|title|yes|String||Title of the information tab.|
|description|no|String||Text displayed under the title.|
|subtexts|no|**[subtexts](#markdown-header-footerfooterInfosubtexts)**[]||Array of subtext configuration objects.|

***

### footer.footerInfo.subtexts
|Name|Required|Type|Default|Description|
|----|-------------|---|-------|------------|
|subtitle|yes|String||Subtitle of the subtext.|
|text|no|String||Text displayed under the subtitle.|

**Beispiel:**

```json
{
    "footerInfo": [
        {
            "title": "Contact",
            "description": "Text under the titel",
            "subtexts": [
                {
                    "subtitle": "Postal address",
                    "text": "Max-Mustermann-Str. 1 <br> 12345 City <br> Germany"
                },
                {
                    "subtitle": "Phone and fax",
                    "text": "Tel: +49 (0) 1234 56789 <br> Fax: +49 (0) 1234 5678910"
                }
            ]
        },
        {
            "title": "Privacy",
            "subtexts": [
                {
                    "subtitle": "Subtitle",
                    "text": "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt..."
                }
            ]
        }
    ]
}
```
***

## login

This module allows the user to login with an OIDC server. The retrieved access token is stored in cookies which can be used by the backend to deliver user-specific data (e.g. layers). Since the cookies are technically required to implement the login functionality, there is not corresponding cookie notice.

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|oidcAuthorizationEndpoint|yes|String||The oidc auth endpoint, e.g. "https://idm.domain.de/auth/realms/REALM/protocol/openid-connect/auth".|
|oidcTokenEndpoint|yes|String||The oidc token endpoint, e.g. "https://idm.domain.de/auth/realms/REALM/protocol/openid-connect/token".|
|oidcClientId|yes|String||The oidc client, e.g. "masterportal" (must be created in your IDM, e.g. keycloak).|
|oidcScope|yes|String||The scope used for oidc, defaults to "profile email openid".|
|oidcRedirectUri|yes|String||The url to redirect the oidc process to - after login.|
|interceptorUrlRegex|yes|String||An regexp pattern that allows to specify urls the oidc token will be attached to.|

Make sure in keycloak the client is configured as follows:

```
Access Type: public
Standard Flow Enabled: ON
Valid Redirect URIs: <PORTAL URL, e.g. https://localhost/*>
Web Origins: <PORTAL HOST, e.g. https://localhost>
```

Specific ports are not allowed ports here. Especially, ports on localhost, e.g. `localhost:9001` will not work with keycloak since ports are not accepted in web origins.

In the section `OpenID Connect Compatibility Modes` activate `Use Refresh Tokens`. In the section `Advanced Settings` set `Proof Key for Code Exchange Code Challenge Method` to `S256`.

**Example:**

```json
{
  "login": {
      "oidcAuthorizationEndpoint": "https://idm.DOMAIN.de/auth/realms/REALM/protocol/openid-connect/auth",
      "oidcTokenEndpoint": "https://idm.DOMAIN.de/auth/realms/REALM/protocol/openid-connect/token",
      "oidcClientId": "masterportal",
      "oidcScope": "profile email openid",
      "oidcRedirectUri": "https://localhost/portal/basic/",
      "interceptorUrlRegex": "https?://localhost.*" // add authorization to all URLs that match the given regex
  }
}
```

## mapInteractions

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|interactionModes|no|Object|`{dragPan: false, altShiftDragRotate: false, pinchRotate: false}`| Interaction settings for the ol default interactions. If not set, the default setting is used.|
|keyboardEventTarget|no|Object/false|`false`|possibility to set the keyboard event target for the ol map e.g keyboardEventTarget: document|

**Example:**

```json
{
    "mapInteractions": {
        "interactionModes": {"dragPan": false, "altShiftDragRotate": true, "pinchRotate": false, "dragZoom": true},
        "keyboardEventTarget": false
    }
}
```

***
## mapMarker

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|pointStyleId|no|String|`"defaultMapMarkerPoint"`|StyleId to refer to a `style.json` point style. If not set, the `img/mapMarker.svg` is used.|
|polygonStyleId|no|String|`"defaultMapMarkerPolygon"`|StyleId to refer to a `style.json` polygon style.|

**Example:**

```json
{
    "mapMarker": {
        "pointStyleId": "customMapMarkerPoint",
        "polygonStyleId": "customMapMarkerPolygon"
    }
}
```

***

## mouseHover

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|minShift|no|Integer|`5`|Minimum mouse position movement required to render a new tooltip; in pixels.|
|numFeaturesToShow|no|Integer|`2`|Maximum amount of element information per tooltip; when exceeded, an information text informs the user of cut content.|
|infoText|no|String|`"(Further objects. Please zoom.)"`|Information text shown when `numFeaturesToShow` is exceeded.|
|highlightOnHover|no|Boolean|false|If hovered features should be highlighted|
|highlightVectorRulesPolygon|no|Object||Specify the fill color and outline color and stroke width for highlighting the polygon features as well as a zoom parameter.|
|highlightVectorRulesPointLine|no|Object||Specify outline color and stroke width for highlighting lines and fill color and scale factor for highlighting points as well as a zoom parameter.|

**Example:**

```json
{
    "mouseHover": {
        "numFeaturesToShow": 2,
        "infoText": "The Info Text",
        "highlightOnHover": true,
        "highlightVectorRulesPolygon": {
            "fill": {
                "color": [255, 255, 255, 0.5]
            },
            "stroke": {
                "width": 4,
                "color": [255, 0, 0, 0.9]
            }
        },
        "highlightVectorRulesPointLine": {
            "stroke": {
                "width": 8,
                "color": [255, 0, 255, 0.9]
            },
            "image": {
                "scale": 2
            }
        }
    },
}
```

### mouseHover.highlightVectorRulesPolygon

Specify the fill color and outline color and stroke width for highlighting the polygon features as well as a zoom level.

|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|fill|no|**[fill](#markdown-header-mousehoverhighlightvectorrulespolygonfill)**||Possible setting: color|false|
|stroke|no|**[stroke](#markdown-header-mousehoverhighlightvectorrulespolygonstroke)**||Possible setting: width|false|

***

### mouseHover.highlightVectorRulesPolygon.fill
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"fill": { "color": [215, 102, 41, 0.9] }
```

***

### mouseHover.highlightVectorRulesPolygon.stroke
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|width|no|Integer|1|Possible setting: width|false|
|color|no|Float[]|[255, 0, 0, 0.9]|Possible setting: color (RGBA)|false|

```json
"stroke": { "width": 4 , "color": [255, 0, 255, 0.9]}
```

***


### mouseHover.highlightVectorRulesPointLine

Specify outline color and stroke width for highlighting lines and fill color and scale factor for highlighting points. Also a zoom level.

|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|fill|no|**[fill](#markdown-header-mousehoverhighlightvectorrulespointlinefill)**||Possible setting: color|false|
|stroke|no|**[stroke](#markdown-header-mousehoverhighlightvectorrulespointlinestroke)**||Possible setting: width|false|
|image|no|**[image](#markdown-header-mousehoverhighlightvectorrulespointlineimage)**||Possible setting: scale|false|

***
### mouseHover.highlightVectorRulesPointLine.fill
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"fill": { "color": [215, 102, 41, 0.9] }
```

***

### mouseHover.highlightVectorRulesPointLine.stroke
|Name|Required|Type|Default|Description|Expert|
|----|-------------|---|-------|------------|------|
|width|no|Integer|1|Possible setting: width|false|
|color|no|Float[]|[255, 255, 255, 0.5]|Possible setting: color (RGBA)|false|

```json
"stroke": { "width": 4 , "color": [255, 0, 255, 0.9]}
```

***

### mouseHover.highlightVectorRulesPointLine.image
|Name|Required|Type|Default|Description|Expert|
|----|--------|----|-------|-----------|------|
|scale|no|Integer|1.5|Possible setting: scale|false|

```json
"image": { "scale": 2}
```

***

## portalLanguage

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|yes|Boolean|`true`|Controls whether a button to switch the portal's language is provided.|
|debug|no|Boolean|`false`|Controls whether debug information regarding translations is logged to the console.|
|languages|yes|Object|`{ de: "deutsch", en: "englisch" }`|Language abbreviations. Please mind that matching locale files must exist.|
|fallbackLanguage|no|String|`"de"`|Fallback language used if contents are not available in the currently selected language.|
|changeLanguageOnStartWhen|no|String[]|`["querystring", "localStorage", "navigator", "htmlTag"]`|Order of user language detection. See [i18next browser language detection documentation](https://github.com/i18next/i18next-browser-languageDetector) for details.|
|loadPath|no|String|`"/locales/{{lng}}/{{ns}}.json"`|Path to load language files from, or a function returning such a path: `function(lngs, namespaces) { return path; }`. `lng` and `ns` are read from the path, if given, as if from a static path. You may also provide a URL like `"https://localhost:9001/locales/{{lng}}/{{ns}}.json"`. See [i18next http backend documentation](https://github.com/i18next/i18next-http-backend) for details.|

**Example:**

```json
{
"portalLanguage": {
        "enabled": true,
        "debug": false,
        "languages": {
            "de": "deutsch",
            "en": "englisch"
        },
        "fallbackLanguage": "de",
        "changeLanguageOnStartWhen": ["querystring", "localStorage", "navigator", "htmlTag"],
        "loadPath": "/locales/{{lng}}/{{ns}}.json"
    }
}
```

***

## quickHelp

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|imgPath|no|String|`"/"`|An absolute or relative path to the folder containing the quick help images.|
|searchbarAllgemeines1|no|String|`"allgemein.png"`|First quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines2|no|String|`"allgemein_2.png"`|Second quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines3|no|String|`"allgemein_3.png"`|Third quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarFlurstueckssuche|no|String|`"allgemein_4.png"`|Quick help image regarding the Searchbar, belonging to the chapter "Parcel search". The image must exist in the `imgPath` folder.|
|aufbau1|no|String|`"themen.png"`|First quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|aufbau2|no|String|`"themen_2.png"`|Second quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|routingTool1|no|String|`"routing_1.png"`|First quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool2|no|String|`"routing_2.png"`|Second quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool3|no|String|`"routing_3.png"`|Third quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool4|no|String|`"routing_4.png"`|Fourth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool5|no|String|`"routing_5.png"`|Fifth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool6|no|String|`"routing_6.png"`|Sixth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool7|no|String|`"routing_7.png"`|Seventh quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool8|no|String|`"routing_8.png"`|Eighth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool9|no|String|`"routing_9.png"`|Nineth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|
|routingTool10|no|String|`"routing_10.png"`|Tenth quick help image regarding the routing tool (RoutingTool). The image must exist in the `imgPath` folder.|

CHANGELOG.md
# Changelog Masterportal
 All important changes in this project are stored in this file.

[Semantic versioning](https://semver.org/spec/v2.0.0.html) is used.

## Unreleased - in development
### __Breaking Changes__

### Added
- The following packages have been added:
    - devDependencies:
        - webpack-bundle-analyzer: 4.9.1
- To manage dependencies of dependencies "overrides" for cesium are added to package.json and create dummy packages for it.
- Issue #1105: Added easting and nothing labels for utm projections in CoordToolkit

### Changed
- The script "npm run build" has been extended. The user can now specify the name of an addon that is excluded from bundle.
- Issue #1064: Print: parameter to configure invisible layer info

### Deprecated

### Removed
- The following packages have been removed:
    - devDependencies:
        - webpack-visualizer-plugin

### Fixed
- prepareFeaturesFor3D is now called after loading
- Issue #1081: Transmit authorization token to services after login
- Issue #1093: Fixed the Wrong CRS in GeoJSON by Download filtered data
- Issue #1099: The package.json is adapted, now the size of the bundle, created with "npm run build" is reduced.

---
## v2.39.0 - 2023-11-01

### Added
- Add roadmap information to the readme file.
- Add additional wfs and wms layer param 'gfiThemeSettings' to provide the possibility of additional gfiTheme settings.
- Issue #1046: Add documentation in the config.json md-files for the `mapView` attribute `startResolution`.

### Changed
- Using the workaround `NODE_OPTIONS=--openssl-legacy-provider` for node 18 is no longer necessary.
- The following NPM packages have been updated:
    - dependencies:
        - @masterportal/masterportalapi: 2.28.0 to 2.29.0 (This also raised ol to version 8.1.0 and olcs to version 2.16.0)
    - devDependencies:
        - esbuild-loader: 3.0.1 to 3.1.0
        - webpack: 4.46.0 to 4.47.0
- Updated readme with contribution and pull request information.

### Removed
- The following NPM packages have been removed:
    - devDependencies:
        - cross-env

### Fixed
- addWMS: Transformation for external services with EPSG:4326
- Issue #886: Remove warning for `Cesium could not be loaded`, if control `button3d` is not configured.
- Issue #1018: The tool selectFeatures can now also be used mobile, as an area can be dragged on the map.
- Issue #1052: VectorStyle: path to an icon in the map or in the legend in style.json may be an icon name, a relative path to icon, a complete url or an svg.
- Issue #1083: fixed the issue.
- Issue #1086: prepush hook runs on linux.
- Issue #1089: When called via the url parameter ?query, no error message is displayed now.
- Fix some vulnerabilities in dependencies
- parametricUrl: show alert again if invalid layerId is used in the url.

---

## v2.38.0 - 2023-10-04
### Changed
- The following NPM packages have been updated:
  - dependencies:
    - @masterportal/masterportalapi: 2.27.0 to 2.28.0 (This also raised ol to version 7.5.2)
- GFI: the gfi in 3D is not sorted reversed

### Fixed
- Custom treetype can now load gdi layer via parametriUrl
- Issue #889: Layer is not shown in the tree anymore
- Issue #1023: Fix layer information for GeoJSON
- Issue #1074: print drawn symbols
- Issue #1075: Routing tool now works as menu entry or tool entry
- AddWMS: WMS with subfolders are now displayed correctly in the topic tree again and can be switched on and off.

---

## v2.37.0 - 2023-09-06
### Added
- New Tool 3D-Modeler has been added to draw/import 3D models into the 3D map.
- Routing
    - Possibility to use *ESRI CH LocationFinder* for geosearch option.
    - Possibility to use *Komoot* for geosearch/reverse geosearch option.
    - Possibility to use *Gazetteer* for geosearch option.
    - Possibility to use *specialWfs* for geosearch option.
    - Possibility to use *Elastic Search* for geosearch option.
    - Possibility to define a bbox value for the BKG geosearch service for the different speed profiles.
- WebGL render pipeline supports now also VectorTile layers.
- Printing is now for 3d supported.
- Added new module `login button` that allows the user to login with an OIDC IDM. The resulting auth token is added to all service requests (can be configured).
- Gitignore
    - added pipeline file for bitbucket
    - added pipeline file for gitlab
- urls in gfiAttributes can now be custimized and shown as html tags (e.g. a, img, iframe) with custom properties
- New shared-component "TableComponent".

### Changed
- The following packages have been updated:
    - dependencies:
        - @masterportal/masterportalapi: 2.23.0 to 2.27.0 (This also raised olcs to version 2.15.0 and ol to 7.5.1)
- Filter:
    - The parameter filterOnMove now only works in combination with "'multiLayerSelector': false" to create a better user experience.
    - The filter configuration in the portal master is adapted to the new syntax.
    - New Parameter `operatorForAttrName` for Snippets to allow the snippet to filter over multiple attrNames with an OR logic.
    - Parameter `input` and `slider` now also work for the snippet type `slider`.

### Fixed
- Issue #776: show only 2D layers in 2D mode and 3D layers in 3D mode
- Issue #1061: Filter: Accept gml:featureMembers occuring with other children in XML response.
- Issue #1068: Print: fixed printing of point layers.
- Issue #1070: show quickHelp icon when search is configured as an object
- Default Tree
    - on category change all layers are shown in the layer tree
    - Config.tree.orderBy sets category

---

## v2.36.1 - 2023-08-04
### Fixed
- WFS: avoid createLegend undefined error

---

## v2.36.0 - 2023-08-02
### Added
- QuickHelp: A description of the route preference `recommended route` has been added.
- 3D:
    - Colored highlighting for 3D Tiles has been added.
    - GFI MapMarker appears now on top of elevations by click.
- Filter:

portal/umweltatlas/config.js
const Config = {
    wfsImgPath: "./resources/img/",
    namedProjections: [
        [
            "EPSG:25832",
            "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs",
        ],
        [
            "EPSG:25833",
            "+title=ETRS89/UTM 33N +proj=utm +zone=33 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs",
        ],
    ],
    footer: {
        urls: [
            {
                bezeichnung: "common:modules.footer.designation",
                url: "https://www.berlin.de/sen/sbw/",
                alias: "Senatsverwaltung für Stadtentwicklung, Bauen und Wohnen",
                alias_mobil: "SenSBW",
            },
        ],
        showVersion: true,
    },
    quickHelp: {
        imgPath: "./resources/img/",
    },
    layerConf: "./resources/services-internet.json",
    restConf: "./resources/rest-services-internet.json",
    styleConf: "./resources/style_v3.json",
    scaleLine: true,
    mouseHover: {
        numFeaturesToShow: 2,
        infoText: "(weitere Objekte. Bitte zoomen.)",
    },
    portalLanguage: {
        enabled: false,
        debug: false,
        languages: {
            de: "Deutsch",
            en: "English",
        },
        fallbackLanguage: "de",
        changeLanguageOnStartWhen: ["querystring", "localStorage", "htmlTag"],
    },
};

// conditional export to make config readable by e2e tests
if (typeof module !== "undefined") {
    module.exports = Config;
}

